### DISCUSSION TITLE
use ThreadPoolExecutor and copy_current_request_context  in flask will appear some bug

### DISCUSSION THREAD

**Discussion by LionSniper** [pallets/flask#5397] [Q&A]
in app.py file: 
```
from flask import Flask
app = Flask(__name__)

@app.route("/")
def hello_world():
    from test2 import start
    start()
    return "<p>Hello, World!</p>"

if __name__ == "__main__":
    app.run(debug=True)
```

in test2 file:
```
from flask import copy_current_request_context
import time
from concurrent.futures import ThreadPoolExecutor
import threading

@copy_current_request_context
def function_a(number):
    print(threading.current_thread().ident)
    time.sleep(1)
    return number

def start():
    with ThreadPoolExecutor(max_workers=5) as executor:
        response = list(executor.map(function_a, [1, 2]))
    return response
```

when i request this api , it will appear "AssertionError: Popped wrong app context. (&lt;flask.ctx.AppContext object at 0x00000167CB79A3C8&gt; instead of &lt;flask.ctx.AppContext object at 0x00000167CB79A668&gt;) // Werkzeug Debugger", how to resolve it?  thanks!

**Comment by davidism** [comment]
Fairly sure this is because `ThreadPoolExecutor` on its own doesn't handle `contextvars`. There's probably something in [`asgiref`](https://github.com/django/asgiref/tree/main/asgiref) that handles this, but I'm not familiar with it enough to point you more than that. Maybe @pgjones knows?
